transport/src/main/java/io/netty/channel/PendingWritesTrackerFactory.java

```diff
@@ -0,0 +1,97 @@
+            return new PendingWritesTracker() {
+                @Override
+                public void incrementPendingOutboundBytes(long bytes) {
+                    pipeline.incrementPendingOutboundBytes(bytes);
+                }
+
+                @Override
+                public void decrementPendingOutboundBytes(long bytes) {
+                    pipeline.decrementPendingOutboundBytes(bytes);
+                }
+
+                @Override
+                public int size(Object msg) {
+                    return pipeline.estimatorHandle().size(msg);
+                }
+            };
+        } else {
+            final MessageSizeEstimator.Handle estimator = ctx.channel().config().getMessageSizeEstimator().newHandle();
+            final ChannelOutboundBuffer buffer = ctx.channel().unsafe().outboundBuffer();
+            if (buffer == null) {
+                return new PendingWritesTracker() {
+                    @Override
+                    public void incrementPendingOutboundBytes(long bytes) {
+                        // noop
+                    }
+
+                    @Override
+                    public void decrementPendingOutboundBytes(long bytes) {
+                        // noop
+                    }
+
+                    @Override
+                    public int size(Object msg) {
+                        return estimator.size(msg);
+                    }
+                };
+            } else {
+                return new PendingWritesTracker() {
+                    @Override
+                    public void incrementPendingOutboundBytes(long bytes) {
+                        // We need to guard against null as channel.unsafe().outboundBuffer() may returned null
+                        // if the channel was already closed when constructing the PendingWriteQueue.
+                        // See https://github.com/netty/netty/issues/3967
+                        if (buffer != null) {
+                            buffer.incrementPendingOutboundBytes(bytes);
+                        }
+                    }
+
+                    @Override
+                    public void decrementPendingOutboundBytes(long bytes) {
+                        // We need to guard against null as channel.unsafe().outboundBuffer() may returned null
+                        // if the channel was already closed when constructing the PendingWriteQueue.
+                        // See https://github.com/netty/netty/issues/3967
+                        if (buffer != null) {
+                            buffer.decrementPendingOutboundBytes(bytes);
+                        }
+                    }
+
+                    @Override
+                    public int size(Object msg) {
+                        return estimator.size(msg);
+                    }
+                };
+            }
+        }
+    }
+
+    public interface PendingWritesTracker extends MessageSizeEstimator.Handle {
```

Comment:
Package private for now please

Reply:
If I make this with package visibility then the `io.netty.handler.ssl.SslHandler.SslHandlerCoalescingBufferQueue` will not be able to access it.

Reply:
maybe annotate with ```@UnstableApi``` then?